hail/python/hail/methods/pca.py

```diff
@@ -207,13 +207,260 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
+    * :math:`U\in\mathbb{R}^{m\times (p+1)b` and :math:`V\in\mathbb{R}^{n\times (p+1)b` are
+      orthonormal matrices (:math:`U^TU = I` and :math:`V^TV = I`)
+    * :math:`\mathrm{span}(V) = \mathcal{K}_p(A^TA, V_0)`
+    * :math:`UR=AV`, hence :math:`\mathrm{span}(U) = \mathcal{K}_p(AA^T, AV_0)`
+    * :math:`V[:, :b] = V_0`
+    * :math:`R\in\mathbb{R}^{b\times b}` is upper triangular
+    where :math:`\mathcal{K}_p(X, Y)` is the block Krylov subspace
+    :math:`\mathrm{span}(Y, XY, \dots, X^pY)`.
+
+    Parameters
+    ----------
+    A_expr
+    V0
+    p
+    compute_U
+
+    Returns
+    -------
+
+    """
+    t = A.block_table
+    A_expr = A.block_expr
+
+    g_list = [V0]
+    G_i = V0
+    k = hl.eval(V0.shape[1])
+
+    for j in range(0, p):
+        info(f"krylov_factorization: Beginning iteration {j+1}/{p}")
+        G_i = t.aggregate(hl.agg.ndarray_sum(A_expr.T @ (A_expr @ G_i)), _localize=False)
+        G_i = hl.nd.qr(G_i)[0]._persist()
+        g_list.append(G_i)
+
+    info("krylov_factorization: Iterations complete. Computing local QR")
+    V0 = hl.nd.hstack(g_list)
+
+    if compute_V:
+        V = hl.nd.qr(V0)[0]._persist()
+        t = t.annotate(AV=A_expr @ V)
+    else:
+        V = hl.nd.qr(V0)[0]
+        t = t.annotate(AV=A_expr @ V)
+        V = None
+
+    if compute_U:
+        temp_file_name = hl.utils.new_temp_file("_krylov_factorization_intermediate", "ht")
+        t = t.checkpoint(temp_file_name)
+        AV_local = t.aggregate(hl.nd.vstack(hl.agg.collect(t.AV)), _localize=False)
+        U, R = hl.nd.qr(AV_local)._persist()
+    else:
+        Rs = t.aggregate(hl.nd.vstack(hl.agg.collect(hl.nd.qr(t.AV)[1])), _localize=False)
+        R = hl.nd.qr(Rs)[1]._persist()
+        U = None
+
+    return KrylovFactorization(U, R, V, k)
+
+
+def _reduced_svd(A: TallSkinnyMatrix, k=10, compute_U=False, iterations=2, iteration_size=None):
+    # Set Parameters
+    q = iterations
+    if iteration_size is None:
+        L = k + 2
+    else:
+        L = iteration_size
+    assert((q + 1) * L >= k)
+    n = A.ncols
+
+    # Generate random matrix G
+    G = hl.nd.zeros((n, L)).map(lambda n: hl.rand_norm(0, 1))
+    G = hl.nd.qr(G)[0]._persist()
+
+    fact = _krylov_factorization(A, G, q, compute_U)
+    info("_reduced_svd: Computing local SVD")
+    return fact.reduced_svd(k)
+
+
+@typecheck(A=oneof(expr_float64, TallSkinnyMatrix),
+           num_moments=int,
+           p=nullable(int),
+           moment_samples=int,
+           block_size=int)
+def _spectral_moments(A, num_moments, p=None, moment_samples=500, block_size=128):
+    if not isinstance(A, TallSkinnyMatrix):
+        check_entry_indexed('_spectral_moments/entry_expr', A)
+        A = _make_tsm_from_call(A, block_size)
+
+    n = A.ncols
+
+    if p is None:
+        p = min(num_moments // 2, 10)
+
+    # TODO: Breaks when moment_samples > n. But in that case, we should just
```

Comment:
Maybe let's add an assert or something for this until it's resolved, or one telling people they ought to just compute eigenvalues directly?

Reply:
Ah, thanks, that's why I left the todo.